File: src/clustr/tools/write/container_power.py
"""
Write tools for LXC container power management.

Same pattern as vm_power.py — separate tools per action, no catch-all.

Annotations:
  - start:    destructiveHint = False
  - shutdown: destructiveHint = False  (graceful)
  - stop:     destructiveHint = True   (force kill)
  - reboot:   destructiveHint = False  (graceful)
"""

from __future__ import annotations

import logging
from typing import Annotated, cast

from mcp.server.fastmcp import FastMCP
from mcp.types import ToolAnnotations
from pydantic import Field

from clustr.proxmox.client import ProxmoxError, get_client, proxmox_post
from clustr.tools import needs_confirm, safe

logger = logging.getLogger(__name__)

_SAFE_WRITE = ToolAnnotations(readOnlyHint=False, destructiveHint=False)
_DESTRUCTIVE = ToolAnnotations(readOnlyHint=False, destructiveHint=True)

_Node = Annotated[str, Field(description="Node name (e.g. 'pve')")]
_CtId = Annotated[int, Field(ge=100, description="Container ID")]
_Confirm = Annotated[
    bool,
    Field(
        description="Must be true to execute this destructive operation. When "
        "false (default), returns a confirmation prompt without acting."
    ),
]


# ---------------------------------------------------------------------------
# Tool implementation
# ---------------------------------------------------------------------------


def _container_power_action(node: str, ctid: int, action: str) -> str:
    actions = {"start", "shutdown", "stop", "reboot"}
    if action not in actions:
        raise ProxmoxError(f"Unknown container power action: {action}")
    return cast(
        str,
        proxmox_post(
            lambda: getattr(get_client().nodes(node).lxc(ctid).status, action).post()
        ),
    )


def _run(node: str, ctid: int, action: str) -> str:
    task_id = _container_power_action(node, ctid, action)
    return (
        f"✅ Container {ctid} on {node}: **{action}** initiated.\n"
        f"Task ID: `{task_id}`\n\n"
        f"Use `get_container_status` to check the current state."
    )


# ---------------------------------------------------------------------------
# Tool registration
# ---------------------------------------------------------------------------


def register(mcp: FastMCP) -> None:
    """Register all container power tools onto the given FastMCP instance."""

    @mcp.tool(
        name="start_container",
        title="Start Container",
        description="Start a stopped LXC container. No effect if already running.",
        annotations=_SAFE_WRITE,
    )
    async def start_container(node: _Node, ctid: _CtId) -> str:
        return await safe("start_container", lambda: _run(node, ctid, "start"))

    @mcp.tool(
        name="shutdown_container",
        title="Shutdown Container (Graceful)",
        description=(
            "Gracefully shut down an LXC container by sending a shutdown signal "
            "to init. Use stop_container if the container is unresponsive."
        ),
        annotations=_SAFE_WRITE,
    )
    async def shutdown_container(node: _Node, ctid: _CtId) -> str:
        return await safe("shutdown_container", lambda: _run(node, ctid, "shutdown"))

    @mcp.tool(
        name="stop_container",
        title="Stop Container (Force)",
        description=(
            "Force-stop an LXC container immediately. "
            "Data loss may occur. Prefer shutdown_container for graceful stop."
        ),
        annotations=_DESTRUCTIVE,
    )
    async def stop_container(
        node: _Node, ctid: _CtId, confirm: _Confirm = False
    ) -> str:
        def _do() -> str:
            if not confirm:
                return needs_confirm("force-stop", f"container {ctid} on {node}")
            return _run(node, ctid, "stop")

        return await safe("stop_container", _do)

    @mcp.tool(
        name="reboot_container",
        title="Reboot Container (Graceful)",
        description="Gracefully reboot a running LXC container.",
        annotations=_SAFE_WRITE,
    )
    async def reboot_container(node: _Node, ctid: _CtId) -> str:
        return await safe("reboot_container", lambda: _run(node, ctid, "reboot"))

File: clustr-ts/src/tools/write/containerPower.ts
/**
 * Write tools for LXC container power management. LXC has no hard "reset", so
 * only stop (force) is destructive.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { proxmoxPost } from "../../proxmox.js";
import { needsConfirm, safe } from "../../safe.js";

const SAFE_WRITE = { readOnlyHint: false, destructiveHint: false } as const;
const DESTRUCTIVE = { readOnlyHint: false, destructiveHint: true } as const;

const node = z.string().describe("Node name (e.g. 'pve')");
const ctid = z.number().int().min(100).describe("Container ID");
const confirm = z
  .boolean()
  .default(false)
  .describe(
    "Must be true to execute this destructive operation. When false (default), " +
      "returns a confirmation prompt without acting.",
  );

async function run(n: string, id: number, action: string): Promise<string> {
  const task = await proxmoxPost(`/nodes/${n}/lxc/${id}/status/${action}`);
  return (
    `✅ Container ${id} on ${n}: **${action}** initiated.\n` +
    `Task ID: \`${String(task)}\`\n\n` +
    "Use `get_container_status` to check the current state."
  );
}

export function register(server: McpServer): void {
  server.registerTool(
    "start_container",
    {
      title: "Start Container",
      description: "Start a stopped LXC container. No effect if already running.",
      inputSchema: { node, ctid },
      annotations: SAFE_WRITE,
    },
    async ({ node: n, ctid: id }) =>
      safe("start_container", () => run(n, id, "start")),
  );

  server.registerTool(
    "shutdown_container",
    {
      title: "Shutdown Container (Graceful)",
      description:
        "Gracefully shut down an LXC container (signals init). Use " +
        "stop_container if it is unresponsive.",
      inputSchema: { node, ctid },
      annotations: SAFE_WRITE,
    },
    async ({ node: n, ctid: id }) =>
      safe("shutdown_container", () => run(n, id, "shutdown")),
  );

  server.registerTool(
    "stop_container",
    {
      title: "Stop Container (Force)",
      description:
        "Force-stop an LXC container immediately. Data loss may occur. Prefer " +
        "shutdown_container for a graceful stop.",
      inputSchema: { node, ctid, confirm },
      annotations: DESTRUCTIVE,
    },
    async ({ node: n, ctid: id, confirm: ok }) =>
      safe("stop_container", async () =>
        ok
          ? run(n, id, "stop")
          : needsConfirm("force-stop", `container ${id} on ${n}`),
      ),
  );

  server.registerTool(
    "reboot_container",
    {
      title: "Reboot Container (Graceful)",
      description: "Gracefully reboot a running LXC container.",
      inputSchema: { node, ctid },
      annotations: SAFE_WRITE,
    },
    async ({ node: n, ctid: id }) =>
      safe("reboot_container", () => run(n, id, "reboot")),
  );
}
